### PR TITLE
Move GitProviders into 'core' module.

### PR DESCRIPTION
**Pull request by mvitz**
These providers are reusable by other plugins.

To make this work they can't depend on any maven classes (e.g.
MojoExecutionException) so I had to introduce an own GitException.

Another change was to refactor the GitDataProvider mechanism from using
inheritance to composition. Therefore the GitProvider interface is now
the main abstraction used by the 'core' module. The GitDataProvider now
just assembles all properties by using a specific instance of the
GitProvider.

Relates to #228.

### CODE REVIEW

File: core/src/main/java/pl/project13/git/api/GitDescribeConfig.java [hunk 1]
```diff
@@ -15,7 +15,7 @@
  * along with git-commit-id-plugin.  If not, see <http://www.gnu.org/licenses/>.
  */
 
-package pl.project13.maven.git;
+package pl.project13.git.api;
```
Comment (ktoso) [inline core/src/main/java/pl/project13/git/api/GitDescribeConfig.java]: good package name choice :+1:

**Comment by TheSnoozer** [comment]
Looks good for me. @ktoso Should I merge? :-)

**Comment by ktoso** [comment]
Looks Good To Me! :-)
Let's merge to keep the momentum going!

Thanks a lot @mvitz for keeping at it!

**Comment by TheSnoozer** [comment]
Note to myself: this was merged into `git-commit-id:modularization-effort` and never made it to the actual repo. Modularisation effort was done in https://github.com/git-commit-id/maven-git-commit-id-plugin/pull/441

**Pull request merged**